File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def helper(self, root, ans):
        if root == None: return []
        q = deque()
        q.append(root)
        left_to_right = True
        while len(q) != 0:
            level = []
            for _ in range(len(q)):
                node = q.popleft()
                level.append(node.val)
                if node.left is not None:
                    q.append(node.left)
                if node.right is not None:
                    q.append(node.right)
            if left_to_right == False:
                level.reverse()
            ans.append(level)
            left_to_right = not left_to_right

    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        ans = []
        self.helper(root,ans)
        return ans
        

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.cpp

class Solution {
public:
    vector<vector<int>> helper(TreeNode* root, vector<vector<int>>& ans) {
        if (root == nullptr)
            return {};

        queue<TreeNode*> q;
        q.push(root);

        bool left_to_right = true;

        while (!q.empty()) {
            vector<int> level;

            int size = q.size();

            for (int i = 0; i < size; i++) {
                TreeNode* node = q.front();
                q.pop();

                level.push_back(node->val);

                if (node->left != nullptr) {
                    q.push(node->left);
                }

                if (node->right != nullptr) {
                    q.push(node->right);
                }
            }

            if (!left_to_right) {
                reverse(level.begin(), level.end());
            }

            ans.push_back(level);

            left_to_right = !left_to_right;
        }

        return ans;
    }

    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;

        helper(root, ans);

        return ans;
    }
};

